api/src/main/java/io/opentelemetry/api/metrics/batch/BatchObserverContext.java

```diff
@@ -0,0 +1,86 @@
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Util class for creating multiple instruments that can be atomically observed in a single
+ * callback.
+ *
+ * <p>This class is equivalent to individually calling {@code observe} on every {@link Instrument},
+ * but has the advantage of all recordings occurring atomically.
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * meter.newBatchObserver(
+ *     context -> {
+ *       DoubleSumBatchObserver timeObserver =
+ *           context.doubleSumObserverBuilder("time").setUnit("s").build();
+ *       LongSumBatchObserver distanceObserver =
+ *           context.longSumObserverBuilder("distance").setUnit("m").build();
+ *
+ *       context.registerCallback(
+ *           result ->
+ *               MyMeasure measure = MyMeasure.getdata();
+ *               result.observe(
+ *                   Labels.of("myKey", "myValue"),
+ *                   distanceObserver.observation(measure.getDistance()),
+ *                   timeObserver.observation(measure.getTime())));
+ *     });
+ * }</pre>
+ */
+@ThreadSafe
+public interface BatchObserverContext extends Instrument {
+
+  /**
+   * Returns a new builder for a {@link DoubleValueBatchObserver}.
+   *
+   * @param name the name of the instrument. Should be a ASCII string with a length no greater than
+   *     255 characters.
+   * @return a new builder for a {@code DoubleValueBatchObserver}.
+   * @throws NullPointerException if {@code name} is null.
+   * @throws IllegalArgumentException if different metric with the same name already registered.
+   * @throws IllegalArgumentException if the {@code name} does not match the requirements.
+   */
+  DoubleValueBatchObserver.Builder doubleValueObserverBuilder(String name);
+
+  /**
+   * Returns a new builder for a {@link LongValueBatchObserver}.
+   *
+   * @param name the name of the instrument. Should be a ASCII string with a length no greater than
+   *     255 characters.
+   * @return a new builder for a {@code LongValueBatchObserver}.
+   * @throws NullPointerException if {@code name} is null.
+   * @throws IllegalArgumentException if different metric with the same name already registered.
+   * @throws IllegalArgumentException if the {@code name} does not match the requirements.
+   */
+  LongValueBatchObserver.Builder longValueObserverBuilder(String name);
+
+  /**
+   * Set a {@link Consumer} function that accepts a {@link BatchObserverContext}. This function will
+   * call {@code observe} to execute the observation of multiple {@link Instrument}s.
+   *
+   * @param resultConsumer Function to consume a {@link BatchObserverContext}.
+   */
+  void registerCallback(Consumer<BatchObserverResult> resultConsumer);
+
+  /** Builder class for {@link BatchObserverContext}. */
+  interface Builder extends Instrument.Builder {
+    @Override
+    BatchObserverContext build();
+  }
+
+  @FunctionalInterface
+  interface BatchObserverResult {
+    void observe(Labels labels, Observation... observations);
```

Comment:
Unnecessary array allocation for the 99% of cases where only one value per observation.

Reply:
Why would it be one value?

The goal of this is to support atomic capture of multiple instrument values, so there should be at least two.

Reply:
I think one would not choose a BatchObserver in the case where there is more than one observation for the same label set?

Reply:
Maybe I'm confused, I thought the goal of [BatchObserver](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#batch-observer) was to capture observations from multiple instruments within a single atomic observation with an identical label set.

Am I misinterpreting something?

Reply:
I think @bogdandrutu is suggesting that maybe we only need one API for both..you could just use the batch API for a single instrument.